### PR TITLE
feat: edk based dbt extension

### PR DESCRIPTION
**Pull request by pandemicsyn**
Temp implementation checklist:

- [x] deploys file-dbt-* scaffolding
- [x] deploys appropriate profile.yml
- [x] change cwd to dbt project dir for execution
- [x] dbt output looks/behaves the same as it does today 
- [x] deploy profiles for the ~4~ 5 dbt transformer variants
  - [x] postgres
  - [x] snowflake
  - [x] redshift
  - [x] bigquery 
  - [x] duckdb (recently added)
- [ ] ~paste able configs for the 4 dbt transfomer variants (redshift, postgres, snowflake, bigquery)~
- [x] staged hub entries over on the hub repo: https://github.com/meltano/hub/pull/841
- [x] pre-invoke `dbt deps`

### CODE REVIEW

**Comment by pandemicsyn** [comment]
@tayloramurphy @pnadolny13 @aaronsteers would one of you mind pulling this down and testing it? I'm slightly suspicious that everything "just worked" - I was expecting this to be the most problematic build.  You should be able to use this config: 

https://github.com/meltano/dbt-ext/blob/320049fd2262d5b8ae6a667fce27f627e9445b4d/README.md

Sample install:

```
(ext-dev) ➜  extdev rm -rf .meltano/utilities/dbt-postgres
(ext-dev) ➜  extdev meltano install utilities dbt-postgres
Installing 1 plugins...
Installing utility 'dbt-postgres'...
Installed utility 'dbt-postgres'
(ext-dev) ➜  extdev tree
.
├── cron-meltano.yml
├── custom
│   └── transform
│       └── models
│           └── tap_gitlab
│               ├── commits_last_7d.sql
│               └── source.yml
├── dbt-meltano.yml
├── meltano.yml
├── plugins
│   ├── extractors
│   └── loaders
├── superset
│   ├── __pycache__
│   │   └── superset_config.cpython-38.pyc
│   ├── superset.db
│   └── superset_config.py
└── transform
    ├── dbt_packages
    ├── logs
    │   └── dbt.log
    ├── models
    │   └── tap_gitlab
    │       ├── commits_last_7d.sql
    │       └── source.yml
    └── profiles

15 directories, 11 files
(ext-dev) ➜  extdev meltano invoke dbt-postgres:initialize
creating dbt profiles directory path=PosixPath('/Users/syn/projects/meltano-projects/extdev/transform/profiles/postgres')
dbt initialized                dbt_ext_type=postgres dbt_profiles_dir=PosixPath('/Users/syn/projects/meltano-projects/extdev/transform/profiles/postgres') dbt_project_dir=PosixPath('/Users/syn/projects/meltano-projects/extdev/transform')
(ext-dev) ➜  extdev tree
.
├── cron-meltano.yml
├── custom
│   └── transform
│       └── models
│           └── tap_gitlab
│               ├── commits_last_7d.sql
│               └── source.yml
├── dbt-meltano.yml
├── meltano.yml
├── plugins
│   ├── extractors
│   └── loaders
├── superset
│   ├── __pycache__
│   │   └── superset_config.cpython-38.pyc
│   ├── superset.db
│   └── superset_config.py
└── transform
    ├── dbt_packages
    ├── dbt_project.yml
    ├── logs
    │   └── dbt.log
    ├── models
    │   └── tap_gitlab
    │       ├── commits_last_7d.sql
    │       └── source.yml
    └── profiles
        └── postgres             <----- postgres profile was deployed
            └── profiles.yml

16 directories, 13 files
```

Sample run:

```
(ext-dev) ➜  extdev meltano --environment=dev invoke dbt-postgres run
2022-09-15T20:59:20.174096Z [info     ] Environment 'dev' is active
Extension executing `dbt deps`...
20:59:23  Running with dbt=1.1.2
20:59:23  Warning: No packages were found in packages.yml


Extension executing `dbt run`...
20:59:24  Running with dbt=1.1.2
20:59:24  Found 1 model, 0 tests, 0 snapshots, 0 analyses, 167 macros, 0 operations, 0 seed files, 2 sources, 0 exposures, 0 metrics
20:59:24
20:59:30  Concurrency: 2 threads (target='dev')
20:59:30
20:59:30  1 of 1 START table model public.commits_last_7d ................................ [RUN]
20:59:31  1 of 1 OK created table model public.commits_last_7d ........................... [SELECT 0 in 1.41s]
20:59:32
20:59:32  Finished running 1 table model in 8.48s.
20:59:32
20:59:32  Completed successfully
20:59:32
20:59:32  Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```

**Review by pnadolny13 — COMMENTED** [review]
@pandemicsyn I was able to install it but it doesnt look like the bundled files are arriving properly after copying the plugin definition into my meltano.yml and running `meltano install utilities dbt-postgres`. So I get an error when initializing:

```
(meltano) Patricks-MacBook-Pro:dbt_ext_testing pnadolny$ meltano invoke dbt-postgres:initialize
2022-09-16T19:56:26.936263Z [info     ] Environment 'dev' is active
Traceback (most recent call last):
  File "/Users/pnadolny/Documents/Git/meltano_project/dbt_ext_testing/dbt_ext_testing/.meltano/utilities/dbt-postgres/venv/lib/python3.9/site-packages/dbt_ext/main.py", line 34, in initialize
    ext.initialize(force)
  File "/Users/pnadolny/Documents/Git/meltano_project/dbt_ext_testing/dbt_ext_testing/.meltano/utilities/dbt-postgres/venv/lib/python3.9/site-packages/dbt_ext/extension.py", line 108, in initialize
    for entry in ir_files("files_dbt_ext.bundle.transform").iterdir():
  File "/Users/pnadolny/.pyenv/versions/3.9.7/lib/python3.9/importlib/resources.py", line 147, in files
    return _common.from_package(_get_package(package))
  File "/Users/pnadolny/.pyenv/versions/3.9.7/lib/python3.9/importlib/_common.py", line 14, in from_package
    return fallback_resources(package.__spec__)
  File "/Users/pnadolny/.pyenv/versions/3.9.7/lib/python3.9/importlib/_common.py", line 18, in fallback_resources
    package_directory = pathlib.Path(spec.origin).parent
  File "/Users/pnadolny/.pyenv/versions/3.9.7/lib/python3.9/pathlib.py", line 1082, in __new__
    self = cls._from_parts(args, init=False)
  File "/Users/pnadolny/.pyenv/versions/3.9.7/lib/python3.9/pathlib.py", line 707, in _from_parts
    drv, root, parts = self._parse_args(args)
  File "/Users/pnadolny/.pyenv/versions/3.9.7/lib/python3.9/pathlib.py", line 691, in _parse_args
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType
initialize failed with uncaught exception, please report to maintainer
```

Screen Shot 2022-09-16 at 4 03 33 PM

I've got a test project set up so I should be able to quickly test moving forward. Let me know if you want me to test again.

Also this reference has a typo in the git url https://github.com/meltano/dbt-ext/blob/320049fd2262d5b8ae6a667fce27f627e9445b4d/README.md?plain=1#L17.

File: README.md
Comment (aaronsteers) [reply README.md:17]: @pnadolny13 - Do you have a branch open in the `squared` repo that I can use for testing?

If not, can we go ahead and modify the `dbt-snowflake` entry or add a new entry for testing?

1. Most of the lockfile can probably stay the same.
2. `pip_url` would get the addition of `git+https://github.com/meltano/dbt-ext.git@feat/working-dbt-ext`
3. We'd move the entry from `transformers` to `utilities` (or create a new one there until `utilities`).

**Comment by pandemicsyn** [comment]
> @pandemicsyn - What do you think of shortening to init instead of initialize as the scaffold-install command?

tl;dr sounds good to me! All for less typing 😁

**Comment by pandemicsyn** [comment]
> @pandemicsyn I was able to install it but it doesnt look like the bundled files are arriving properly after copying the plugin definition into my meltano.yml and running meltano install utilities dbt-postgres.

@pnadolny13 Gah, sorry. Pretty sure I know whats going on. Let me poke around a bit more.

File: README.md
Comment (pnadolny13) [reply README.md]: @aaronsteers I dont have a branch for that but I can make one and post it back here.

**Comment by pandemicsyn** [comment]
@pnadolny13 ok should work now. I'd used 3.8 so far testing and turns out importlib behaves different in +3.9 🙄

**Comment by pnadolny13** [comment]
> @pnadolny13 ok should work now. I'd used 3.8 so far testing and turns out importlib behaves different in +3.9 🙄

@pandemicsyn yeah it worked for me. One thing I noticed was the dbt deps before each run including when I run dbt deps 😄 .

File: README.md
Comment (pnadolny13) [reply README.md]: @aaronsteers heres the draft PR in squared using the extension. Everything worked for me so far.

**Comment by pandemicsyn** [comment]
> @pandemicsyn yeah it worked for me. One thing I noticed was the dbt deps before each run including when I run dbt deps 😄 .

Yep, right now, it runs it every time - we could limit it to just having it run before specific commands (dbt run and dbt test ?). I thought someone mentioned having deps always exec via the pre-invoke hook, but I probably just misunderstood.
  
@pnadolny13 @aaronsteers any prefs on having it automagically invoked before specific commands vs not automagically invoking it at all?

**Comment by pandemicsyn** [comment]
fyi - threw a sample dbt-snowflake yaml in the readme. I don't have a good way of testing it but should work unless I introduced some typo's.

File: README.md
Comment (aaronsteers) [reply README.md]: Great! Thanks, @pnadolny13 !

cc @pandemicsyn

**Comment by aaronsteers** [comment]
> @pnadolny13 @aaronsteers any prefs on having it automagically invoked before specific commands vs not automagically invoking it at all?

I don't have strong opinions or deep knowlege on this part. I know Meltano was invoking them as a prehook but I don't know if that was only before specific commands.

@tayloramurphy - do you know?

**Comment by tayloramurphy** [comment]
@aaronsteers @pnadolny13 our best practice was that we always ran `dbt deps` because it was a relatively low-risk operation, especially if your dependencies were pinned. It could slow run time a bit because it would have to fetch the packages, but that saved time because you wouldn't have a weird error because something was missing. 

Ideally we would be able to tell if we _needed_ to run dbt deps before a command, but it'd take some effort to understand what the criteria for triggering it or not would be. Interestingly, on dbt Cloud they automatically run `dbt clean` prior to `dbt deps`. We did this at GitLab as well https://gitlab.com/gitlab-data/analytics/-/blob/master/Makefile#L10 and https://gitlab.com/gitlab-data/analytics/-/blob/master/Makefile#L127

I think it's reasonable to run it before every command (with the exception of `dbt deps` if that isn't too much of a change). We could also add `dbt clean` in there as well (searching dbt slack shows that's a recommended path for a lot of folks) but it could get annoying on _every_ run especially if you have to re-compile everything every time. 

Let's do this:
* add dbt clean to the pre-invoke along with deps
* skip pre-invoke if doing clean or deps separately
* document the pre-invoke we're doing and how to turn it off if desired

does that work?

**Comment by pnadolny13** [comment]
@tayloramurphy +1 to what you said. Skipping the deps if you're actually running deps since its redundant and having this pre-invoke feature on by default but with a setting that I can toggle if I want to skip it for speed.

**Comment by pandemicsyn** [comment]
> Let's do this:

> add dbt clean to the pre-invoke along with deps
> skip pre-invoke if doing clean or deps separately
> document the pre-invoke we're doing and how to turn it off if desired
> does that work?

@tayloramurphy  🤘 sounds like a solid plan. Will send a long an update here in a bit.

**Comment by pandemicsyn** [comment]
fyi - just need to land

-  https://github.com/meltano/edk/pull/25

...and then update this PR to match whatever pattern we settle on there.

**Comment by pandemicsyn** [comment]
Update - the pre-invoke is conditional now. Will always be skipped when running `clean` or `deps` , and can be completely turned off via `meltano config dbt-postgres set skip-pre-invoke` or on-demand for big invocations with `DBT_EXT_SKIP_PRE_INVOKE=true meltano run dbt-postgres:test dbt-postgres:freshness dbt-postgres:run ...`

**Review by aaronsteers — APPROVED** [review]
Approved from my side. But should wait for at least one other approval from Data Team and one from Engineering.

**Comment by pandemicsyn** [comment]
fyi - i tested this pr with the MDS in a box repo that @tayloramurphy shared and looks like it worked as expected. 

https://github.com/pandemicsyn/nba-monte-carlo/tree/florian/test-dbt-ext <-- fork & branch of the `MDS in a box` utilizing dbt-ext.

**Comment by pandemicsyn** [comment]
Chatted with @aaronsteers about this earlier in the week. This got some extra testing since I trialed it with MDS - so going to go ahead and merge it without further approval's so as not to remain blocked. 

There's a PR to add this to the hub waiting in the wings, and that PR will likely see additional testing from others.

**Pull request merged**
